### ISSUE TITLE
Term Weeks for Lent 2016 misaligned

### ISSUE THREAD

**Issue by waiwaing** [camdram/camdram#332]
Full Term for Lent 2016 runs Tue 12 Jan - Fri 11 Mar. Therefore, 

Term Week 1: Thurs 14 Jan - Wed 20 Jan
Term Week 2: Thurs 21 Jan - Wed 27 Jan
etc.

The corresponding Theatre weeks are therefore:

Theatre Week 1: Sun 17 Jan - Sat 23 Jan
Theatre Week 2: Sun 24 Jan - Sat 30 Jan

However, Camdram's 2016 Lent Term assigns week 0 to Theatre Week 1.

**Comment by waiwaing** [comment]
I think the error is at [line 43 in TimePeriodUpdatesCommand](https://github.com/camdram/camdram/blob/master/src/Acts/CamdramBackendBundle/Command/TimePeriodsUpdateCommand.php#L43), which contains a off-by-one bug. 

This happens when the calculation date (7 Oct, 17 Jan, 25 Apr) falls on a Sunday, e.g. Lent 2016.

For example, see Michaelmas 2012, when 7 Oct was on a Sunday. Rewinding to Sunday doesn't change the date, but as term started on 2 Oct, it will calculate that as week 0 instead of week 1. To correct this, a full week should be subtracted when the calculation date is on a Sunday, rather than 0 days.

The fix is to remove lines 43 and 45, remove the relevant database entries for Lent 2016 weeks, and rerun the script.

**Comment by hoyes** [comment]
Cheers Wai-Wai. The intention of "rewindToSunday" was to subtract 0 <= x <= 6 days from the current date, so I just changed the 3 hard-coded "latest start of week 0" dates back one day. Will re-run the script once this code hits Tiberius

**Comment by alexbrett** [comment]
Reopening this - I've just had to manually fix in the DB the entries for Easter 2016. Camdram had showed week 0 as beginning Sunday 24th April, when it's actually Sunday 17th April (as full term starts Tuesday 19th).

While I've resolved it on the live system for now, the script presumably needs fixing, otherwise it may end up overwriting my manual DB fix the next time it's run?

**Comment by waiwaing** [comment]
I've just done some quick checking and there appears to be no single date for [line 31](https://github.com/camdram/camdram/blob/master/src/Acts/CamdramBackendBundle/Command/TimePeriodsUpdateCommand.php#L31) which will guarantee a correct result given the dates of Full Term (Ordinances, Chapter II, Section 10).  Looking into a nice and elegant solution.

**Comment by waiwaing** [comment]
So here's what I believe to be the case:

- The span between the earliest Full Easter Term can start and the latest it can start is greater than one week. This is also true for Michaelmas and Lent as it happens.
- The date which Full Easter Term starts is not dependent on the date of Easter.
- The date which Full Easter Term starts is not consistently the nth Tuesday of April (nor the last nor second to last Tuesday), nor is the Sunday on which Theatre Week 0 starts similarly the nth Sunday of April (nor likewise).

I thus do not believe that the date Full Easter Term starts can be algorithmically determined (and it is likely that Michaelmas/Lent are the same), and certainly not easily algorithmically determined.

**Comment by philosophicles** [comment]
This conclusion was what I had originally believed based on some conversation I had around the time I matriculated. So I suspect you are right. I suggest in the short term, we lapse back to a V1-style approach where we manually add rows as required to show the dates of near-future terms. 

In the longer term we could look at screenscraping the official UoC term dates page: this could probably be a standalone code job offered up under a separate repository, which we simply use. Or, I've just discovered that Google Docs has a basic scraping feature built-in, which we might be able to use somehow: https://eagereyes.org/data/scrape-tables-using-google-docs. I've just tried this in a Google Doc on my personal account and it works fine with the simple cell function `=IMPORTHTML("https://www.cam.ac.uk/about-the-university/term-dates-and-calendars","table",1)`.
